### PR TITLE
Builtin keccak

### PR DESCRIPTION
**Pull request by jkktom**
Draft PR for Builtin - Keccak256
Done
[x] add padding
[x] applied keccak as CairoKeccak
[x] test result from the web identical (verified)
[x] LE, BE (Little Endian, Big Endian) way of Keccak Created as Rust VM does
[ ] test for LE, BE Keccak

Side Note : in the Cairo Repo, there's "starknet_keccak" in Rust code. Basically doing the same thing. location : https://github.com/cairo/crates/cairo-lang-starknet/src/contract.rs

### CODE REVIEW

**Comment by jkktom** [comment]
Dear @rodrigo-pino , This PR is ready for review.
Keccak256 works for suitable inputs with appropriate paddings applied. All tested with success.
All the test inputs in test codes verified from Web result : https://emn178.github.io/online-tools/keccak_256.html 
Same inputs, Same results.

Still contains Println and Printf lines for verification and debug. This will be commented out after reveiw.
Thanks.

**Comment by jkktom** [comment]
Resolves issue #130

**Comment by jmjac** [comment]
Spend some time today trying to figure out how this should work and it seems the keccak builtin does the keccak-f1600 permutation to the 8 felt 200bit inputs and writes 8 felt 200bit outputs.

File: pkg/vm/builtins/keccak.go [hunk 1]
```diff
@@ -0,0 +1,257 @@
...
+		if !value.Known() {
+			return fmt.Errorf("cannot infer value: input value at offset %d is unknown", startOffset+i)
+		}
+		v, err := value.FieldElement()
+		if err != nil {
+			return fmt.Errorf("Keccak input has to be felt")
+		}
+		var out [32]byte
+		fp.LittleEndian.PutElement(&out, *v)
+		copy(data[i*25:i*25+25], out[:25]) //25*8 = 200bits
```
Comment (rodrigo-pino) [inline pkg/vm/builtins/keccak.go:46]: 25 * 8 = 200 bytes ?

File: pkg/vm/builtins/keccak.go [hunk 2]
```diff
@@ -0,0 +1,257 @@
...
+	var output [200]byte
+	for i := 0; i < 25; i++ {
+		v := binary.LittleEndian.AppendUint64(nil, out[i])
+		copy(output[i*8:i*8+8], v)
+	}
+
+	for i := 0; i < inputCellsPerKeccak; i++ {
+		var bytes [32]byte
+		copy(bytes[:], output[i*25:i*25+25])
+		//Only 200 bits so always fits, no need to check error
```
Comment (rodrigo-pino) [inline pkg/vm/builtins/keccak.go:65]: Only 200 bits, or 200 bytes? I am unsure if it should be bits, or bytes. Since data structures are in bytes but comments are in bits. Shouldn't all be bits?

File: pkg/vm/builtins/keccak_math.go [hunk 3]
```diff
@@ -0,0 +1,411 @@
+package builtins
+
+import "math/bits"
+
+// Slighly modified from: https://github.com/golang/crypto/blob/master/sha3/keccakf.go
```
Comment (rodrigo-pino) [inline pkg/vm/builtins/keccak_math.go:5]: What modifications we required?

File: pkg/vm/builtins/keccak.go [hunk 4]
```diff
@@ -0,0 +1,257 @@
...
+		}
+	}
+	return nil
+}
+
+func (k *Keccak) String() string {
+	return KeccakName
+}
+
+const (
```
Comment (rodrigo-pino) [inline pkg/vm/builtins/keccak.go:80]: Are we using the code below this line?

**Review by rodrigo-pino — COMMENTED** [review]
@jkktom @jmjac good work! Just a couple of questions, and remember we need to add an integration test too. Just like we have one for the other builtins.

File: pkg/vm/builtins/keccak.go (hunk 1 above)
Comment (jmjac) [reply pkg/vm/builtins/keccak.go:46]: It's 200 bits. We read 25 bytes from each felt which gives us 200bits per felt. 1600bits in total from the 8 inputs

File: pkg/vm/builtins/keccak.go (hunk 2 above)
Comment (jmjac) [reply pkg/vm/builtins/keccak.go:65]: Maybe rewording it to `This is 25*8 bits which is smaller than max felt 252 bits so no need to check the error` would make it clearer?

Comment (rodrigo-pino) [reply pkg/vm/builtins/keccak.go:65]: Yes, we are talking about bits, but the data slice for example holds 200 bytes, shouldn't it hold 25 bytes instead?

File: pkg/vm/builtins/keccak_math.go (hunk 3 above)
Comment (jmjac) [reply pkg/vm/builtins/keccak_math.go:5]: As I remember I had to change it to work on values instead of pointers during testing. Looking at it now it works the just fine with the original code so I'll change it back to that for easier compatibility if go's keccakf1600 ever gets updated.

File: pkg/vm/builtins/keccak.go (hunk 4 above)
Comment (jmjac) [reply pkg/vm/builtins/keccak.go:80]: Not as I'm aware though @jkktom may know if it maybe used in the future for cairo 1

File: pkg/vm/builtins/keccak.go (hunk 2 above)
Comment (jmjac) [reply pkg/vm/builtins/keccak.go]: There are two arrays here. The first output holds 200 bytes because keccakf1600 works on [25]uint64 which is also 200 bytes. Then it does the reverse of the first loop and converts each 25bytes to a felt which gives us in total 200bytes converted to 8 felts where each felt is 25bytes=200bits.

File: pkg/vm/builtins/keccak.go (hunk 4 above)
Comment (jkktom) [reply pkg/vm/builtins/keccak.go]: DIscussed with Jacob, to be deleted as it's not necessary.

Comment (rodrigo-pino) [reply pkg/vm/builtins/keccak.go:86]: Let's put this code in `pkg/utils/keccak.go`, to have it separated from the builtin

File: pkg/vm/builtins/keccak_test.go
Comment (rodrigo-pino) [inline pkg/vm/builtins/keccak_test.go:1]: Let's add a test as well

Comment (jmjac) [reply pkg/vm/builtins/keccak_test.go:1]: Moved and commented out debug statements to not clutter the test output

**Review by rodrigo-pino — APPROVED** [review]
Amazing! Thanks @jkktom and @jmjac
